drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,305 @@
+            for (j, byte) in v.to_le_bytes().iter().enumerate() {
+                out[i * 4 + j] = *byte;
+            }
+        }
+        out
+    }
+
+    fn read_reg(&self, reg: *mut u32) -> u32 {
+        unsafe { reg.read_volatile() }
+    }
+
+    fn write_reg(&self, reg: *mut u32, v: u32) {
+        unsafe { reg.write_volatile(v) };
+    }
+
+    /// Wait until the FPGA is idle
+    fn wait_fpga_busy(&self) {
+        loop {
+            let status = self.read_reg(reg::SPISR);
+            if (status & 1) == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            sleep_for(1);
+        }
+    }
+
+    /// Wait until a word is available in the RX buffer
+    fn wait_fpga_rx(&self) {
+        for i in 0.. {
+            let status = self.read_reg(reg::SPISR);
+            // Wait for the
+            const RX_EMPTY_BIT: u32 = 1 << 6;
+            if status & RX_EMPTY_BIT == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            // Initial busy-loop for faster response
+            if i >= 32 {
+                sleep_for(1);
+            }
+        }
+    }
+
+    /// Wait until the SPI flash is idle
+    fn wait_flash_busy(&self, t: Trace) {
+        // Wait for the busy flag to be unset
+        while (self.read_flash_status() & 1) != 0 {
+            ringbuf_entry!(t);
+            sleep_for(1);
+        }
+    }
+
+    fn read_flash_status(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_1);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    fn read_flash_status_3(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_3);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    fn flash_write_enable(&self) {
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::WRITE_ENABLE);
+        self.wait_fpga_busy();
+    }
+
+    fn clear_fifos(&self) {
+        self.write_reg(reg::SPICR, 0x8080);
+    }
+
+    fn flash_sector_erase(&mut self, addr: u32) {
+        self.flash_write_enable();
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, addr);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::BLOCK_ERASE_64KB_4B);
+
+        // Wait for the busy flag to be unset
+        self.wait_flash_busy(Trace::SectorEraseBusy);
+    }
+
+    fn flash_read<'a, B: idol_runtime::BufWriter<'a>>(
```

Comment:
To avoid generating multiple copies of this function, I'd strongly suggest switching this to use `&mut dyn BufWriter` instead of a type parameter.

Reply:
done!